pythonfuzz/corpus.py

```diff
@@ -104,181 +493,19 @@ def mutate(self, buf):
-                for k in range(4):
-                    res[pos+k] = numpy.uint8(res[pos+k] + v[k])
-            elif x == 10:
-                # Add/subtract from a uint64.
-                if len(res) < 8:
-                    i -= 1
-                    continue
-                pos = self._rand(len(res) - 7)
-                v = numpy.uint64(self._rand(35) + 1)
-                if bool(random.getrandbits(1)):
-                    v = numpy.uint64(0) - v
-                if bool(random.getrandbits(1)):
-                    v = struct.pack('>Q', v)
-                else:
-                    v = struct.pack('<Q', v)
-                for k in range(8):
-                    res[pos+k] = numpy.uint8(res[pos+k] + v[k])
-            elif x == 11:
-                # Replace a byte with an interesting value.
-                if len(res) == 0:
-                    i -= 1
-                    continue
-                pos = self._rand(len(res))
-                res[pos] = numpy.uint8(INTERESTING8[self._rand(len(INTERESTING8))])
-            elif x == 12:
-                # Replace an uint16 with an interesting value.
-                if len(res) < 2:
-                    i -= 1
-                    continue
-                pos = self._rand(len(res) - 1)
-                v = numpy.uint16(INTERESTING16[self._rand(len(INTERESTING16))])
-                if bool(random.getrandbits(1)):
-                    v = struct.pack('>H', v)
-                else:
-                    v = struct.pack('<H', v)
-                res[pos] = numpy.uint8(v[0])
-                res[pos+1] = numpy.uint8(v[1])
-            elif x == 13:
-                # Replace an uint32 with an interesting value.
-                if len(res) < 4:
-                    i -= 1
-                    continue
-                pos = self._rand(len(res) - 3)
-                v = numpy.uint32(INTERESTING32[self._rand(len(INTERESTING32))])
-                if bool(random.getrandbits(1)):
-                    v = struct.pack('>I', v)
-                else:
-                    v = struct.pack('<I', v)
-                for k in range(4):
-                    res[pos+k] = numpy.uint8(v[k])
-            elif x == 14:
-                # Replace an ascii digit with another digit.
-                digits = []
-                for k in range(len(res)):
-                    if ord('0') <= res[k] <= ord('9'):
-                        digits.append(k)
-                if len(digits) == 0:
-                    i -= 1
-                    continue
-                pos = self._rand(len(digits))
-                was = res[digits[pos]]
-                now = was
-                while was == now:
-                    now = self._rand(10) + ord('0')
-                res[digits[pos]] = now
+
+            # Select a mutator from those we can apply
+            # We'll try up to 20 times, but if we don't find a
+            # suitable mutator after that, we'll just give up.
+            for n in range(20):
```

Comment:
why change to 20 instead of the nm? `nm = self._rand_exp()` was a magic number used both in go-fuzz and afl/libfuzzer and it worked well. I tried other numbers and it usually affected the speed of the fuzzer significantly

Reply:
You're confusing the two loops, I think.

The outer loop (line 495 in the new changes, just above those change) selects the number of mutations to run. Before this change, it was actually the 'number of allocators we'll try to run', because if the mutator said 'sorry, I'm not appropriate' (by decrementing the iteration count) it wouldn't actually affect the number of mutations run.

In this new version, when the mutator says it's not appropriate (by returning None at line 504), we want to try the mutation sequence again. This is what the inner loop of 20 attempts does. 20 was just picked from the air as seeming like a good number.

This means that the number of mutators that we plan to add, as determined in by nm, will be the number we generate (unless we are exceptionally unlucky and repeatedly select inapprorpriate mutations within the inner loop - which can happen if you've filtered the mutations that are used to those that are more likely to complain).

This is a change in behaviour from the original - instead of just ignoring mutations that are unavailable, the code now tries to meet the number that were requested. I chose to follow what appeared to be the intent of the code (that the number of mutators we plan to apply will be the number that we aim for, and that we retry rather than drop those attempts when they fail).

Reply:
In case it's not clear, the inner loop is bounded so that we don't get stuck continually retrying. I originally was lazy and used `while True` and then when testing the filtering, generated a filter that left only a mutator that worked on a non-empty input, but the string always started out as empty when no dictionary or corpus was supplied, so it ran forever.